--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -2573,7 +2573,7 @@
  * other way round. 0 to 1 flip needs to happen via erases.
  */
 static int spi_nor_octal_dtr_write(struct spi_nor *nor, loff_t to, size_t len,
-				   const u8 *buf)
+				   const u8 *buf, u32 page_size)
 {
 	u8 *tmp_buf;
 	size_t bytes_written;
@@ -2583,11 +2583,11 @@
 	if (IS_ALIGNED(to, 2) && IS_ALIGNED(len, 2))
 		return spi_nor_write_data(nor, to, len, buf);
 
-	tmp_buf = kmalloc(nor->params->page_size, GFP_KERNEL);
+	tmp_buf = kmalloc(page_size, GFP_KERNEL);
 	if (!tmp_buf)
 		return -ENOMEM;
 
-	memset(tmp_buf, 0xff, nor->params->page_size);
+	memset(tmp_buf, 0xff, page_size);
 
 	start = round_down(to, 2);
 	end = round_up(to + len, 2);
@@ -2772,7 +2772,7 @@
 
 		if (nor->write_proto == SNOR_PROTO_8_8_8_DTR)
 			ret = spi_nor_octal_dtr_write(nor, addr, page_remain,
-						      buf + i);
+						      buf + i, page_size);
 		else
 			ret = spi_nor_write_data(nor, addr, page_remain,
 						 buf + i);
