--- a/sound/soc/xilinx/xlnx_formatter_pcm.c
+++ b/sound/soc/xilinx/xlnx_formatter_pcm.c
@@ -292,8 +292,7 @@
 {
 	u32 val;
 	void __iomem *reg;
-	struct device *dev = arg;
-	struct xlnx_pcm_drv_data *adata = dev_get_drvdata(dev);
+	struct xlnx_pcm_drv_data *adata = arg;
 
 	reg = adata->mmio + XLNX_MM2S_OFFSET + XLNX_AUD_STS;
 	val = ioread32(reg);
@@ -311,8 +310,7 @@
 {
 	u32 val;
 	void __iomem *reg;
-	struct device *dev = arg;
-	struct xlnx_pcm_drv_data *adata = dev_get_drvdata(dev);
+	struct xlnx_pcm_drv_data *adata = arg;
 
 	reg = adata->mmio + XLNX_S2MM_OFFSET + XLNX_AUD_STS;
 	val = ioread32(reg);
@@ -667,7 +665,7 @@
 
 	ret = devm_request_irq(dev, aud_drv_data->mm2s_irq,
 			       xlnx_mm2s_irq_handler, 0,
-			       "xlnx_formatter_pcm_mm2s_irq", dev);
+			       "xlnx_formatter_pcm_mm2s_irq", aud_drv_data);
 	if (ret) {
 		dev_err(dev, "xlnx audio mm2s irq request failed\n");
 		return ret;
@@ -717,7 +715,7 @@
 	ret = devm_request_irq(dev, aud_drv_data->s2mm_irq,
 			       xlnx_s2mm_irq_handler, 0,
 			       "xlnx_formatter_pcm_s2mm_irq",
-			       dev);
+			       aud_drv_data);
 	if (ret) {
 		dev_err(dev, "xlnx audio s2mm irq request failed\n");
 		return ret;
