diff --git a/drivers/uio/uio_core.c b/drivers/uio/uio_core.c
--- a/drivers/uio/uio_core.c
+++ b/drivers/uio/uio_core.c
@@ -1051,6 +1051,11 @@
 err_request_irq:
 	uio_dev_del_attributes(idev);
 err_uio_dev_add_attributes:
+	mutex_lock(&idev->info_lock);
+	idev->info = NULL;
+	mutex_unlock(&idev->info_lock);
+	wake_up_interruptible(&idev->wait);
+	kill_fasync(&idev->async_queue, SIGIO, POLL_HUP);
 	device_del(&idev->dev);
 err_device_create:
 	uio_free_minor(idev->minor);
